### PR TITLE
Generic implementations for FromJSONKey and ToJSONKey

### LINKED ISSUE #710: Generic ToJSONKey and FromJSONKey
There seems to be no easy way to write common instances of `ToJSONKey`  and `FromJSONKey`. In the following answer I use `Show` and `Read` but that still takes some typing and knowledge about the library.

https://stackoverflow.com/questions/56609201/how-to-use-a-sum-type-as-a-key-in-map-when-using-aeson/56610450#56610450

We should be able to derive this with `Generic`, defining a `gToJSONKeyText` and `gFromJSONKeyTextParser`. Related to issue #463 which suggests this solution as an intermediate step for a harder problem.

### PR DESCRIPTION
**Pull request by Lysxia**
Fixes #710

### CODE REVIEW

**Comment by Lysxia** [comment]
I don't understand the CI failure. Seems unrelated.

**Comment by bergmark** [comment]
That's just the hlint step, so don't worry about it.

**Comment by bergmark** [comment]
Thanks! Looks like this also resolves the first step of #463?

Might be nice to have some full fledged example in the test suite as well, but no blocker for me to merge this.

I'll look into the hlint issue before merging this, to see if it's just this PR causing it or not.

**Comment by Lysxia** [comment]
> Thanks! Looks like this also resolves the first step of #463?

Yes it does!

File: Data/Aeson/Types/ToJSON.hs [hunk 1]
```diff
@@ -515,6 +529,28 @@ contramapToJSONKeyFunction h x = case x of
     ToJSONKeyText  f g -> ToJSONKeyText (f . h) (g . h)
     ToJSONKeyValue f g -> ToJSONKeyValue (f . h) (g . h)
 
+-- 'toJSONKey' for 'Generic' types.
+-- These types must be sums of nullary constructors, whose names will be used
```
Comment (phadej) [inline Data/Aeson/Types/ToJSON.hs]: Deriving is supported for enumeration types, i.e. the sums of nullary constructors. The names of constructors will be used as keys for JSON objects.

File: Data/Aeson/Types/Class.hs [hunk 2]
```diff
@@ -58,6 +58,11 @@ module Data.Aeson.Types.Class
     , fromJSONKeyCoerce
     , coerceFromJSONKeyFunction
     , mapFromJSONKeyFunction
+    -- ** Generic keys
+    , GToJSONKey()
+    , gToJSONKey
```
Comment (phadej) [inline Data/Aeson/Types/Class.hs]: `genericToJSONKey` to to be consistent with `genericToJSON` etc.

**Comment by Lysxia** [comment]
thanks @phadej I incorporated your suggestions.

**Comment by Lysxia** [comment]
Good idea, I'll do that in the next few days.

**Comment by Lysxia** [comment]
For `Nullary` which has only three values, well, we can easily enumerate them, so a simple unit test gives full input coverage (as in this last updated commit).

**Comment by phadej** [comment]
@Lysxia except when you add another case, etc. E.g. when Generics mechanism is expanded, there would be simple property to copy...

**Comment by Lysxia** [comment]
We'll see if and when someone adds such functionality. Until then (which might well be forever), a roundtrip test is redundant with the existing unit tests.

**Comment by bergmark** [comment]
Would probably be nice to have that test as well but not a blocker, I think. I didn't get time to fix travis+hlint before going away over the weekend so I'll merge this now.

**Pull request merged**